### PR TITLE
Migrate to aws-sdk-v2

### PR DESCRIPTION
**Pull request by AlexVulaj**
This PR migrates away all use of the old go aws sdk to the new v2 sdk.

Addresses [OSD-17454](https://issues.redhat.com//browse/OSD-17454)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,10 +6,18 @@ require (
 	cloud.google.com/go/compute v1.19.3
 	github.com/PagerDuty/go-pagerduty v1.5.1
 	github.com/andygrunwald/go-jira v1.16.0
-	github.com/aws/aws-sdk-go v1.44.110
```
Comment (AlexVulaj) [inline go.mod:9]: 👋🏻

File: pkg/provider/aws/iam.go [hunk 2]
```diff
@@ -25,17 +27,19 @@ func CreateIAMUserAndAttachPolicy(awsClient Client, username, policyArn *string)
...
 		}
 		return false, err
 	}
 	return true, nil
 }
 
+func Foo() int {
+	return 1
+}
+
```
Comment (mjlshen) [inline pkg/provider/aws/iam.go:42]: typo?

File: cmd/account/console.go [hunk 3]
```diff
@@ -168,8 +166,8 @@ func (o *consoleOptions) run() error {
 	consoleURL, err := aws.RequestSignInToken(
 		awsClient,
 		&o.consoleDuration,
-		awsSdk.String(sessionName),
-		awsSdk.String(targetRoleArn.String()),
+		&sessionName,
+		pointer.String(targetRoleArn.String()),
```
Comment (mjlshen) [inline cmd/account/console.go:170]: It's technically the same thing, but I'm wondering if it makes more sense for this to be `awsSdk.String` where `awsSdk` is now `github.com/aws/aws-sdk-go-v2/aws`

**Review by mjlshen — CHANGES_REQUESTED** [review]

File: pkg/provider/aws/client.go [hunk 4]
```diff
@@ -166,344 +159,326 @@ func addProxyConfigToSessionOptConfig(config *aws.Config) {
...
 	if _, err := awsClient.GetCallerIdentity(nil); err != nil {
-		if aerr, ok := err.(awserr.Error); ok {
-			switch aerr.Code() {
-			case "InvalidClientTokenId":
-				return nil, fmt.Errorf("`aws sts get-caller-identity --profile %s --region %s` failed - "+
-					"ensure the credentials in the profile are valid and the region is in the AWS partition you expect", profile, region)
-			default:
-				return nil, err
-			}
+		var iite *stsTypes.InvalidIdentityTokenException
```
Comment (mjlshen) [inline pkg/provider/aws/client.go:212]: InvalidIdentityTokenException != InvalidClientTokenId

File: pkg/provider/aws/iam.go (hunk 2 above)
Comment (AlexVulaj) [reply pkg/provider/aws/iam.go:42]: 😆 this was me testing something with my local config. good catch.

File: pkg/provider/aws/client.go (hunk 4 above)
Comment (AlexVulaj) [reply pkg/provider/aws/client.go:212]: The new sts SDK doesn't seem to have an error type of "InvalidClientTokenId", and this was the only error I could find relating to tokens at all so I assumed it was what they renamed it to. The doc for InvalidIdentityTokenException reads:
```
// The web identity token that was passed could not be validated by Amazon Web
// Services. Get a new identity token from the identity provider and then retry the
// request.
```

Any thoughts?

File: pkg/osdCloud/aws.go [hunk 5]
```diff
@@ -335,12 +336,13 @@ func (a *AwsCluster) GetAZs() []string {
 	return a.AZs
 }
 
-func (a *AwsCluster) GetAllVirtualMachines(region string) ([]VirtualMachine, error) {
+func (a *AwsCluster) GetAllVirtualMachines(string) ([]VirtualMachine, error) {
```
[the remark this answers is not in the document]
Comment (AlexVulaj) [reply pkg/osdCloud/aws.go:339]: There's an interface in pkg/osdCloud/interfaces that AwsCluster is trying to satisfy. The region is needed for GCP clusters, so I decided to leave it in the interface and just make it unused (explicitly) for the AWS impl.

File: cmd/account/console.go (hunk 3 above)
Comment (AlexVulaj) [reply cmd/account/console.go:170]: the v2 sdk got rid of that package entirely, so I just made these all `pointer.<>` everywhere applicable.

Comment (AlexVulaj) [reply cmd/account/console.go:170]: Sorry, my first comment misread this.

I'm okay with either, but my preference was to use the standard lib for an operation like this. To compare the functions' source code:

```
AWS SDK V2:
func String(v string) *string {
	return ptr.String(v)
}
// calls down to...
func String(v string) *string {
	return &v
}
```

```
Std Lib:
func String(s string) *string {
	return &s
}
```

So really the same thing is going on under the hood. I'll note that if you (or anyone else 😄 ) feel strongly one way or the other, it'd be extremely easy to do a quick find/replace for `pointer.` through this PR, just let me know!

Comment (mjlshen) [reply cmd/account/console.go:170]: Yeah, if it was in the stdlib, I'd be all for it haha. More that it's been changed to use `"k8s.io/utils/pointer"`?

File: pkg/provider/aws/client.go (hunk 4 above)
Comment (mjlshen) [reply pkg/provider/aws/client.go:210]: Yeah, AWS doesn't have available types for all of their errors, in this case the two errors are not the same. `InvalidClientTokenId` is a common error to STS (https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html#API_GetCallerIdentity_Errors) and can happen during `aws sts get-caller-identity` if the AWS Access Key ID is invalid, but `InvalidIdentityToken` specifically only happens during `aws sts assume-role-with-web-identity` https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html#API_AssumeRoleWithWebIdentity_Errors

File: cmd/account/console.go (hunk 3 above)
Comment (AlexVulaj) [reply cmd/account/console.go:170]: This is a miss by me actually - for some reason I thought this was the `golang.org/x/tools/go/pointer` package 🤦🏻 

Let me revert all of these to use the aws sdk.

File: pkg/provider/aws/client.go (hunk 4 above)
Comment (AlexVulaj) [reply pkg/provider/aws/client.go:210]: Thanks for the insight. In this case, I'm going to drop the special handling of this particular alert and return the error through from the AWS SDK.

File: pkg/provider/aws/client.go [hunk 6]
```diff
@@ -166,344 +157,319 @@ func addProxyConfigToSessionOptConfig(config *aws.Config) {
...
-		if aerr, ok := err.(awserr.Error); ok {
-			switch aerr.Code() {
-			case "InvalidClientTokenId":
-				return nil, fmt.Errorf("`aws sts get-caller-identity --profile %s --region %s` failed - "+
-					"ensure the credentials in the profile are valid and the region is in the AWS partition you expect", profile, region)
-			default:
-				return nil, err
-			}
-		}
+		return nil, fmt.Errorf("error getting caller identity: %w", err)
```
Comment (mjlshen) [inline pkg/provider/aws/client.go:208]: Emphasis that this changed, personally I'm ok with it, but so other reviewers are aware

**Review by mjlshen — APPROVED** [review]
/approve

File: pkg/provider/aws/client.go (hunk 6 above)
Comment (AlexVulaj) [reply pkg/provider/aws/client.go:208]: For context - https://github.com/openshift/osdctl/pull/411#discussion_r1268674566

**Review by mjlshen — APPROVED** [review]
/lgtm

**Pull request merged**